crates/shared/src/event_handling.rs

```diff
@@ -109,45 +115,87 @@ pub struct AlloyEventRetriever<T>(pub T);
 
     async fn get_events_by_block_hash(&self, block_hash: H256) -> Result<Vec<Self::Event>> {
-        self.0
-            .get_events()
-            .at_block_hash(block_hash.into_alloy())
-            .query()
+        let filter = self.0.filter().at_block_hash(block_hash.into_alloy());
+        let events = self
+            .0
+            .provider()
+            .get_logs(&filter)
             .await
-            .map_err(anyhow::Error::from)
+            .map_err(anyhow::Error::from)?
+            .into_iter()
+            .filter_map(|log| {
+                let event = T::Event::decode_log(&log.inner).ok()?;
+                Some((event.data, log))
+            })
+            .collect();
+
+        Ok(events)
     }
 
+    // Unfortunately, alloy's `watch_logs` does not support pagination yet, so it
+    // is implemented manually here
     async fn get_events_by_block_range(
         &self,
         block_range: &RangeInclusive<u64>,
     ) -> Result<EventStream<Self::Event>> {
-        let stream = self
-            .0
-            .get_events()
-            .from_block(*block_range.start())
-            .to_block(*block_range.end())
-            .watch()
-            .await
-            .map_err(anyhow::Error::from)?
-            .into_stream()
-            .map_err(anyhow::Error::from);
+        const CHUNK_SIZE: usize = 500;
+
+        let start = *block_range.start();
+        let end = *block_range.end();
+        let provider = self.0.provider().clone();
+        let base_filter = self.0.filter();
+
+        let stream = futures::stream::iter((start..=end).step_by(CHUNK_SIZE))
+            .then(move |chunk_start| {
+                let provider = provider.clone();
+                let chunk_end = std::cmp::min(chunk_start + (CHUNK_SIZE as u64) - 1, end);
+                let filter = base_filter
+                    .clone()
+                    .from_block(chunk_start)
+                    .to_block(chunk_end);
+
+                async move {
+                    let logs = provider.get_logs(&filter).await.context(format!(
+                        "unable to get logs for blocks range {}-{}",
+                        chunk_start, chunk_end
+                    ))?;
+
+                    let events: Result<Vec<_>> = logs
+                        .into_iter()
+                        .map(|log| {
+                            let event = T::Event::decode_log(&log.inner)
+                                .context(format!("unable to parse log: {:?}", log))?;
+                            Ok((event.data, log))
+                        })
+                        .collect();
+
+                    events
+                }
+            })
+            .map(|chunk_result| {
+                futures::stream::iter(match chunk_result {
+                    Ok(events) => events.into_iter().map(Ok).collect::<Vec<_>>(),
+                    Err(e) => vec![Err(e)],
+                })
+            })
+            .flatten();
```

Comment:
Can we replace `.map().flatten()` with `.flat_map()` here? Not sure if that's a thing for streams.